WebDAV sensor

jkeljo opened the pull request:
## Description:
This sensor monitors a WebDAV share. Its state is simply a boolean
of whether the share is available or not. It has a `files` attribute
that lists all files in the share.


**Related issue (if applicable):** n/a

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8038

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: webdav
    name: file_server
    host: URL_OF_SHARE
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jkeljo committed:
WebDAV sensor

This sensor monitors a WebDAV share. Its state is simply a boolean
of whether the share is available or not. It has a `files` attribute
that lists all files in the share.

fabaff commented:
> Its state is simply a boolean of whether the share is available or not.

If there are only two states then it should be a binary sensor.

fabaff commented on homeassistant/components/sensor/webdav.py:
```diff
@@ -0,0 +1,113 @@
+
+        from webdav3.client import Client
+
+        self._client = Client({
```
Set up the client during `setup_platform()` and pass it to `WebDAVSensor`. Make the platform setup fail if the credentials are wrong.

OttoWinter commented on homeassistant/components/sensor/webdav.py:
```diff
@@ -0,0 +1,113 @@
+    @property
+    def unique_id(self):
+        """Return the URL for the share."""
+        return self._get_url("")
```
Unique IDs should be truly unique. For example if the user changes the path of this sensor, it will have another unique ID and thus an entity_id with `_2` added at the end. See https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements for requirements.

jkeljo replied:
If the user changes the URL, the sensor is now monitoring a different share. Doesn't that make it a different sensor?

MartinHjelmare commented on homeassistant/components/sensor/webdav.py:
```diff
@@ -0,0 +1,113 @@
+            _LOGGER.debug(
+                "Could not open share %s. Message: %s",
+                self.unique_id,
+                str(exception))
```
The exception argument already has a str magic method.

MartinHjelmare commented on homeassistant/components/sensor/webdav.py:
```diff
@@ -0,0 +1,113 @@
+        return self._available
+
+    @property
+    def state_attributes(self):
```
This should be `device_state_attributes`.

jkeljo committed:
- `sensor` -> `binary_sensor`
- `state_attributes` -> `device_state_attributes`
- remove unnecessary `str`
- try to connect in `setup_platform`

jkeljo committed:
Fix listing now that it is calling Client.get_url

jkeljo committed:
Update requirements_all after move to binary sensor

jkeljo committed:
Fix doc URL

dgomes commented on homeassistant/components/binary_sensor/webdav.py:
```diff
@@ -0,0 +1,126 @@
+    from webdav3.client import Client
+    from webdav3.exceptions import WebDavException
+
+    name = config.get(CONF_NAME)
```
config[CONF_NAME] as CONF_NAME is required

dgomes commented on homeassistant/components/binary_sensor/webdav.py:
```diff
@@ -0,0 +1,126 @@
+    from webdav3.exceptions import WebDavException
+
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
```
as with CONF_NAME

dgomes commented on homeassistant/components/binary_sensor/webdav.py:
```diff
@@ -0,0 +1,126 @@
+        return self._name
+
+    @property
+    def unique_id(self):
```
Really should be left out ... it's by no mean unique

jkeljo replied:
You're the second person to push back on this identifier, and I really don't understand it. As I read the wiki, the unique ID must be unique within the platform (i.e. `sensor.webdav`). The URL of the share is the identity of the thing being observed, just like a device serial number or MAC address (examples from the unique ID docs) identifies a device. I could enforce that people cannot configure multiple sensors for the same URL if you guys want, but I haven't seen other platforms prevent the same device from being referenced twice in configuration, even though they use device serial numbers for unique IDs. What am I missing?

dgomes replied:
The thing with unique_id's is that they will live forever in HA (even after you remove the configuration). MAC, EUI, etc aren't user configurable, and that's what's  set them apart. 

Why don't you add a friendly_name configuration parameter instead?

dgomes commented on homeassistant/components/binary_sensor/webdav.py:
```diff
@@ -0,0 +1,126 @@
+    def device_state_attributes(self):
+        """Return the list of files on the share."""
+        return {
+            "files": self._files,
```
I'm concerned this list can become very big and blow up the UI... can we have a limit on the number of files here at least ? say 10 ?

dgomes replied:
OK, you described a use case that is by no means related to this PR.

This binary_sensor is about connectivity (see your DEVICE_CLASS). It should be just about availability of the webdav server an nothing more.

What you wan't is an entity that feeds you files into your photo frame, right ? I think you could model your use case through a **camera**, but I need to understand better your use-case, can you describe in more detail the end result you expect ?

dgomes replied:
Very long thread :)

IMHO I would split Depict from the Samsung Frame TV.

For Depict, you really just need a camera, you pass the camera URL to Depict and that's it. Don't call it camera.depict, call it camera.webdav. It would be nice to have such camera showing up family pictures in the middle of the interface :)

For samsung frame tv... definitely a media_player, ultimately that's what you're doing...

The issue now is the playlist.... for the camera.webdav, the playlist can be part of the component (that is something IMHO fits into the scope of the "camera"... but don't take my word for it). Another option would be to create a new component just for the purpose of the playlist's...

I think we should take baby steps and learn in the process, why don't you start with the camera.webdav ? I would use it :)

balloob commented:
We should not merge this . This is clearly an integration built for a very specific purpose, which does not fit with the general theme of Home Assistant integrations.

A webdav integration would be a component that fires events whenever a file has changed. We should not abuse our state machine to transfer data for automations.
